src/models/train_model.py

```diff
@@ -0,0 +1,205 @@
+                    self.parameters["LGBM"]["learning_rate"]["min"],
+                    self.parameters["LGBM"]["learning_rate"]["max"],
+                    log=True,
+                )
+                n_estimators = trial.suggest_int(
+                    "LGBM_n_estimators",
+                    self.parameters["LGBM"]["n_estimators"]["min"],
+                    self.parameters["LGBM"]["n_estimators"]["max"],
+                )
+                model = LGBMRegressor(
+                    num_leaves=num_leaves,
+                    max_depth=max_depth,
+                    learning_rate=learning_rate,
+                    n_estimators=n_estimators,
+                    random_state=1,
+                    verbose=-1,
+                )
+            elif model_name == "CatBoost":
+                max_depth = trial.suggest_int(
+                    "CB_max_depth", self.parameters["CB"]["max_depth"]["min"], self.parameters["CB"]["max_depth"]["max"]
+                )
+                learning_rate = trial.suggest_float(
+                    "CB_learning_rate",
+                    self.parameters["CB"]["learning_rate"]["min"],
+                    self.parameters["CB"]["learning_rate"]["max"],
+                    log=True,
+                )
+                n_estimators = trial.suggest_int(
+                    "CB_n_estimators",
+                    self.parameters["CB"]["n_estimators"]["min"],
+                    self.parameters["CB"]["n_estimators"]["max"],
+                )
+                model = CatBoostRegressor(
+                    max_depth=max_depth,
+                    learning_rate=learning_rate,
+                    n_estimators=n_estimators,
+                    random_seed=1,
+                    verbose=False,
+                )
+
+            score = cross_val_score(
+                model, x_train, y_train, n_jobs=-1, cv=cv_splitter, scoring="neg_mean_absolute_error"
+            )
+            cv_mae = score.mean()
+            return cv_mae
+
+        sampler = optuna.samplers.TPESampler(seed=1)
+        study = optuna.create_study(direction="maximize", sampler=sampler)
+        study.optimize(objective, n_trials=self.optim_iters, n_jobs=-1)
+        self.model_best_params = study.best_params
+
+        model = create_model_based_on_best_params(self.model_best_params)
+        model.fit(x_train, y_train)
+        logging.info(f"Model trained with following params: {self.model_best_params}")
+        self.model = model
+        params = str(self.model_best_params)
+        hashed_params = hashlib.md5(params.encode()).hexdigest()
+        today = datetime.today().strftime("%Y%m%d%H%M%S")
+        self.model_name = f"{today}_{hashed_params}.pkl"
+
+        def save_model():
+            logging.info(f"Saving model to the file")
+
+            joblib.dump(self.model, self.model_path / self.model_name)
+            logging.info(f"Saved model as {self.model_name}")
+
+        save_model()
```

Comment:
this function should have argument, instead of taking self parameters. when executed we dont know how joblib is created